### ISSUE TITLE
Error on Bech32m format introduced in BIP 350 - invalid bech32 checksum variant found Bech32 when Bech32m was expected

### ISSUE THREAD

**Issue by lukasdll** [rust-bitcoin/rust-bitcoincore-rpc#287]
JsonRpc(Json(Error("invalid bech32 checksum variant found Bech32 when Bech32m was expected", line: 1, column: 77318)))

This error is related to an incorrect Bech32 checksum variant in the JSON-RPC response. The error message indicates that a Bech32 checksum variant was found, but a Bech32m checksum variant was expected.

This issue is most likely caused by an inconsistency between the Bitcoin node (or other service) and the Bitcoin library Bitcoincore_rpc. 

The Bech32m format was introduced in BIP 350 to address some limitations of the original Bech32 format and is used for native segregated witness (SegWit) v1+ addresses.

The transaction below is an example that should allow to reproduce the error, it is one of the new Wasabi transactions they introduced in their last update.

https://mempool.space/tx/371a5a5c3f4344e8f6d49395bfa3f76fe38e5a25db10640a5523c0377a825015

**Comment by apoelstra** [comment]
Can you provide more information? What function from this library were you calling? Is the library outputting data with the wrong checksum or trying to read data with the wrong checksum?

**Comment by lukasdll** [comment]
Absolutely. I can provide a minimal code to replicate the error: Error: JSON-RPC error: JSON decode error: invalid bech32 checksum variant found Bech32 when Bech32m was expected at line 1 column 77318

**cargo.toml**
`[package]
name = "bitcoincore-rpc-example"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bitcoincore-rpc = "0.16.0"
bitcoin = "0.30.0"`

**main.rs**

`use bitcoincore_rpc::{Auth, Client, RpcApi};

fn main() {
    // Hard-code the transaction hash
    let txid = "371a5a5c3f4344e8f6d49395bfa3f76fe38e5a25db10640a5523c0377a825015".parse().unwrap();
    
    // Replace with the Bitcoin node IP, PORT, and credentials if any
    let rpc_auth = Auth::UserPass("username".to_string(), "password".to_string());
    let rpc_client = Client::new("http://<IP>:<PORT>", rpc_auth).unwrap();

    // Get the input information and print it
    match get_input_info(&rpc_client, &txid) {
        Ok(inputs) => {
            println!("inputs len: {}", inputs.len());
            for (idx, input) in inputs.iter().enumerate() {
                println!("Input #{}:", idx);
                println!("  Address: {}", input.0);
                println!("  Value: {} satoshis", input.1);
            }
        }
        Err(e) => eprintln!("Error: {}", e),
    }
}

fn get_input_info(
    rpc_client: &Client,
    txid: &bitcoincore_rpc::bitcoin::Txid,
) -> Result<Vec<(String, f64)>, Box<dyn std::error::Error>> {
    // Get the raw transaction

    // Iterate through transaction inputs and collect input information
    let mut inputs_info = Vec::new();
    let tx = rpc_client.get_raw_transaction(txid, None)?;
    let inputs = tx.input.clone();
    for input in inputs {
        let prev_tx = rpc_client.get_raw_transaction_info(&input.previous_output.txid, None)?;
        let value = prev_tx.vout[input.previous_output.vout as usize].value.to_btc();
        let prev_output = rpc_client.get_tx_out(&input.previous_output.txid, input.previous_output.vout, None)?;
        let tx_out =  prev_output.clone().unwrap();
        let address = tx_out.script_pub_key.addresses.as_ref().and_then(|addrs| addrs.get(0)).unwrap().to_string();
        println!("{} BTC - {}", value, address);
        inputs_info.push((address.to_owned().to_string(), value));
    }

    // Return the input information
    Ok(inputs_info)
}
`

**Comment by lukasdll** [comment]
Upon investigating the root of the problem, it becomes clear that bitcoincore_rpc is not responsible for decoding addresses; instead, it relies on rust_bitcoin (specifically, src/util/address.rs). It might be worthwhile for bitcoincore_rpc to update the dependency version, or if that doesn't resolve the issue, the error may need to be reported to the rust_bitcoin maintainer.

Your feedback on the above would be appreciated Andrew Poelstra

**Comment by apoelstra** [comment]
>the error may need to be reported to the rust_bitcoin maintainer.

I am the rust-bitcoin maintainer :). I will deal with it upstream once I understand the issue. Thanks very much for the specific reproduction!

>I managed to make the code work, but using the method get_raw_transaction instead of get_raw_transaction_info.

This is surprising to me -- both methods should ultimately hand off to rust-bitcoin, which shouldn't have any idea what RPC the request is coming from, and fail in the same way.

**Comment by lukasdll** [comment]
> I am the rust-bitcoin maintainer :). I will deal with it upstream once I understand the issue. Thanks very much for the specific reproduction!

All my respect for your work honestly. I'm extremely grateful to people like you maintaining open source projects. 

> So I changed your code to call gettxout on every input from the previous transactions rather than just the spent one, and that failed because the addresses field was blank, since this field was deprecated and removed 2 years ago in Bitcoin Core 22.

The client is connecting to a Bitcoin Core node, so I checked the version and it is still Bitcoin Core 21, which is probably why I could use the address field. I presume backwards compatibility has certain limitation, and I should probably update that node :)

Your suspicions are perfectly right, and accurate.

**Comment by apoelstra** [comment]
Ok, thanks! I'll try with a Core 21 node then -- I don't think Core would've ever gotten bech32m and bech32 mixed up, so now I'm back to thinking the bug is with rust-bitcoin.

**Comment by lukasdll** [comment]
Are you sure that you couldn't reproduce the error with using get_raw_transaction_info() with the node v 0.21?
I don't think there is anything special in my bitcoin.conf:

```
daemon=1
txindex=1
rpcuser=<username>
rpcpassword=<password>
rpctimeout=50
rpcallowip=<ip/range>
rpcbint=<ip/port>
listen=1
server=1
rpcport=8332
```

**Comment by apoelstra** [comment]
Are you saying that the second code, which you described as "corrected" and that you "got it to work", is actually code that exbhibits the bug? :)

I will try it later today.

**Comment by lukasdll** [comment]
This is the exact code that allows me to reproduce the error. 
Proof :  https://imgur.com/a/fLOFUxy

```
use bitcoincore_rpc::{Auth, Client, RpcApi};

fn main() {
    // Hard-code the transaction hash
    let txid = "371a5a5c3f4344e8f6d49395bfa3f76fe38e5a25db10640a5523c0377a825015".parse().unwrap();

    // Replace IP to correspond with the node endpoint, and set username, and password or make rpc_auth = Auth::None;
    let rpc_auth = Auth::UserPass(<USERNAME>, <PASSWORD>);
    let rpc_client = Client::new("http://<IP>:8332", rpc_auth).unwrap();

    // Get the input information and print it
    let input = get_input_info(&rpc_client, &txid, 0).unwrap();
    println!("  Value: {} BTC", input);
}

fn get_input_info(
    rpc_client: &Client,
    txid: &bitcoincore_rpc::bitcoin::Txid,
    vin: usize
) -> Option<usize> {
    let tx_result = rpc_client.get_raw_transaction(txid, None); // Pass 1 to get a JSON object with transaction details
    match tx_result {
        Ok(tx) => {
            let prev_tx = rpc_client.get_raw_transaction_info(&tx.input[vin].previous_output.txid, None).unwrap();
            let prev_output = prev_tx.blocktime;
            prev_output
        },
        Err(_r) => {
            return None
        }
    }
}
```

And those are my dependencies:

```
[package]
name = "bitcoincore-rpc-example"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bitcoincore-rpc = "0.16.0"
bitcoin = "0.29.0"
```

**Comment by apoelstra** [comment]
Your proof notwithstanding, I cannot reproduce this even with 0.21, using exactly the code you just pasted. It works fine on my end.

**Comment by lukasdll** [comment]
That is, extremely weird. Do I need to checksum the Bitcoin Core node? I'm pretty sure I did not mess with that at all though.
The server is just a raw Ubuntu 22.04 server.

**Comment by lukasdll** [comment]
Could it be due to the length of the transaction, that gets truncated at some point? That transaction is particularly large.

**Comment by apoelstra** [comment]
Maaaaybe. That seems really unlikely -- there is only a tiny tiny chance that a malformed address would appear to be bech32m when it was supposed to be bech32, or vice-versa.

What I suspect is that somehow the node is really really old and doesn't support bech32m. But I dunno, no explanation really seems to make sense to me.

**Comment by lukasdll** [comment]
I could perhaps parse all transactions in a block, and do stats on how many cause an error. If all of them, or only those with certain characteristics. I could also sync a new Bitcoin Core node, up to date, or with a specific version, and observe whether calling the same code causes the error or not. Let me know what would you recommend to pin point the source of the problem.

**Comment by apoelstra** [comment]
I think you should use `tcpdump -A` to find the exact data that the bitcoind is sending over RPC prior to the crash. Then we can at least tell whether the problem is bitcoind's fault or not.
